### ISSUE TITLE
--exclude artist <artist> album <album> excludes all of artist

### ISSUE THREAD

**Issue by ryan-sg** [joshkunz/ashuffle#112]
I have a rather large list of --excludes i've built in a shell script.  I've noticed that excluding via artist + album specification doesn't seem to actually work right -- instead, it just excludes the entire artist.

I noticed when excluding an artist + album combo, and comparing the "Picking random songs out of a pool of..." values from before and after.  Excluding a couple albums from an artist makes the pool value drop by the number that would be met with simply ``--exclude artist <artist>`` by itself.

should I be specifying these differently?  at present I have worked around it by specifying just ``--exclude album <album>``  but this only works when the album doesn't match a different artist as well.

On another note, being able to specify excludes in a file that ashuffle reads from would be helpful, i'm sure at some point i'm going to overflow the max command length with this shellscript :grin:

**Comment by joshkunz** [comment]
Looking at the code, I think you're right that it's considering an exclusion to match if any field matches. In fact, it looks like I [even wrote a test](https://github.com/joshkunz/ashuffle/blob/30e7e247a537d481955c0437b4af4d2b14af1205/t/rule_test.cc#L85-L86) for this behavior. However, this A) seems like not a useful thing for `--exclude` to do when given multiple tags and B) does not match what the [docs say](https://github.com/joshkunz/ashuffle#patterns). I'm going to call this a bug and fix it in a minor revision.

Luckily this should be an easy fix. Just need to tweak the test/condition.

**Comment by joshkunz** [comment]
I filed #113 for the excludes-in-a-file FR.

**Comment by joshkunz** [comment]
Fixed in `v3.12.17`.

**Comment by ryan-sg** [comment]
hmm, it seems to still be happening here, I just build recent checkout from git:

```
21:30 ryan@radio1:~$ ashuffle --exclude artist "april wine" album "live at the el mocambo"
Picking random songs out of a pool of 20870.
^C
21:30 ryan@radio1:~$ ashuffle --exclude album "live at the el mocambo"                     
Picking random songs out of a pool of 21058.
^C
21:30 ryan@radio1:~$ ashuffle --exclude artist "april wine"
Picking random songs out of a pool of 20870.
^C
21:30 ryan@radio1:~$ ashuffle --version
ashuffle version: v3.12.7
```

**Comment by joshkunz** [comment]
Hmm. That's no good. New theory: There's a bug in the parsing logic. I think it may be creating a new "rule" for each group of tag pattern matches. I'll see if I can confirm.

**Comment by joshkunz** [comment]
Indeed, it looks like there was a bug where each pattern in an exclusion rule would become a separate rule. I even had tests to check this silly behavior! A lot of the tests were done during the C->C++ migration so I was just trying to get them to pass. I messed up fixing this one I guess.

Anyways, PR #116 has a fix, and I added an integration test, so I have a pretty high confidence it actually works this time.
